lib/grandpa/network.go

```diff
@@ -174,39 +174,43 @@ func (s *Service) sendMessage(msg GrandpaMessage) error {
 	return nil
 }
 
-func (s *Service) sendNeighbourMessage(interval time.Duration) {
+// notifyNeighbor will gossip a NeighbourMessage every 5 minutes, however we reset the ticker
+// whenever a finalisation occur meaning that a neighbour message already was sent by s.initiateRound()
+func (s *Service) notifyNeighbor(interval time.Duration) {
 	t := time.NewTicker(interval)
 	defer t.Stop()
+
 	for {
 		select {
 		case <-s.ctx.Done():
 			return
 		case <-t.C:
-			if s.neighbourMessage == nil {
-				continue
-			}
-		case info, ok := <-s.finalisedCh:
-			if !ok {
-				// channel was closed
-				return
-			}
-
-			s.neighbourMessage = &NeighbourMessage{
+			s.roundLock.Lock()
+			nm := &NeighbourMessage{
 				Version: 1,
-				Round:   info.Round,
-				SetID:   info.SetID,
-				Number:  uint32(info.Header.Number),
+				Round:   s.state.round,
+				SetID:   s.state.setID,
+				Number:  uint32(s.head.Number),
 			}
-		}
+			s.roundLock.Unlock()
 
-		cm, err := s.neighbourMessage.ToConsensusMessage()
-		if err != nil {
-			logger.Warnf("failed to convert NeighbourMessage to network message: %s", err)
-			continue
+			s.sendNeighborMessage(nm)
+		case <-s.finalisedCh:
```

Comment:
Maybe we can end this goroutine and spawn a new one every time with supplied `round`, `setID` and `head` parameters when we receive something on `finalisedCh`.  That's the more event driven way.  There may be some overlap of messages, but I would assume that's expected.  Or we could explicitly cancel any existing goroutines running `notifyNeighbour`.